### ISSUE TITLE
false positive `unsupported-base` for self-referential generic base

### ISSUE THREAD

**Issue by jorenham** [astral-sh/ty#1800]
### Summary

There are several examples of this in [optype](https://github.com/jorenham/optype), and this is a (stripped down version of) one of them ([src](https://github.com/jorenham/optype/blob/2d03b31e0ce61c6fa827a9753ea131e2aada527a/optype/copy.py#L31-L43)):

```py
from typing import Protocol, override, Self

class CanCopy[T](Protocol):
    def __copy__(self, /) -> T: ...

class CanCopySelf(CanCopy["CanCopySelf"], Protocol):
    @override
    def __copy__(self, /) -> Self: ...
```

```
Unsupported class base with type `<class 'CanCopy[CanCopySelf]'> | <class 'CanCopy[CanCopySelf]'>` (unsupported-base) [Ln 6, Col 19]
```

which runs just fine at runtime:

```pycon
>>> from typing import Protocol, override, Self
... 
... class CanCopy[T](Protocol):
...     def __copy__(self, /) -> T: ...
... 
... class CanCopySelf(CanCopy["CanCopySelf"], Protocol):
...     @override
...     def __copy__(self, /) -> Self: ...
...     
>>> # ¯\_(ツ)_/¯
```

### Version

ty 0.0.1-alpha.32

**Comment by AlexWaygood** [comment]
This is probably https://github.com/astral-sh/ty/issues/1732. I'll check tomorrow if it repros prior to https://github.com/astral-sh/ruff/commit/2c0c5ff4e7a0f347e81bdcdb6e38180dcee0688b landing (if not, it's almost certainly #1732)

**Comment by AlexWaygood** [comment]
This is a bug in ty; we should support this (we shouldn't "need" to infer a union type there at all). But with regards to

> which runs just fine at runtime

note that the error code was `unsupported-base` rather than `invalid-base`. `invalid-base` is about things that will cause runtime errors; `unsupported-base` is specifically about patterns that will not cause errors at runtime but which ty cannot support:

- https://docs.astral.sh/ty/reference/rules/#invalid-base
- https://docs.astral.sh/ty/reference/rules/#unsupported-base

**Comment by carljm** [comment]
Will be fixed by https://github.com/astral-sh/ruff/pull/21909
